Libplanet/Net/Swarm.BlockSync.cs

```diff
@@ -16,436 +20,513 @@ public partial class Swarm<T>
+                }
+
+                if (!complete
+                    || workspace.Tip == BlockChain.Tip
+                    || cancellationToken.IsCancellationRequested)
+                {
+                    _logger.Debug(
+                        $"{nameof(SyncBlocksAsync)}() is aborted. Complete? {complete}; " +
+                        "delete the temporary working chain ({TId}: #{TIndex} {THash}), " +
+                        "and make the existing chain ({EId}: #{EIndex} {EHash}) remains.",
+                        wId,
+                        workspace.Tip.Index,
+                        workspace.Tip.Hash,
+                        BlockChain.Id,
+                        BlockChain.Tip.Index,
+                        BlockChain.Tip.Hash
+                    );
+                }
+                else
+                {
+                    _logger.Debug(
+                        $"{nameof(SyncBlocksAsync)} finished; " +
+                        "replace the existing chain ({0}: {1}) with " +
+                        "the working chain ({2}: {3}).",
+                        BlockChain.Id,
+                        BlockChain.Tip,
+                        wId,
+                        workspace.Tip
+                    );
+
+                    renderSwap = BlockChain.Swap(workspace, render: render);
+                }
+
+                foreach (Guid chainId in chainIds)
                 {
-                    _store.DeleteChainId(workspaceId);
+                    if (!chainId.Equals(BlockChain.Id))
+                    {
+                        _logger.Verbose("Delete an unused chain: {ChainId}", chainId);
+                        wStore.DeleteChainId(chainId);
+                    }
                 }
 
-                throw;
+                _logger.Verbose("Remaining chains: {@ChainIds}", wStore.ListChainIds());
             }
-            finally
+
+            return renderSwap;
+        }
+#pragma warning restore MEN003
+
+        private void ExecuteActions(
+            BlockChain<T> workspace,
+            Block<T> branchpoint,
+            IProgress<PreloadState> progress,
+            CancellationToken cancellationToken)
+        {
+            if (workspace.Renderers.Any())
             {
-                const string msg =
-                    "{SessionId}: " + fname +
-                    "() completed (chain ID: {ChainId}, tip: #{TipIndex} {TipHash}).";
-                _logger.Debug(
-                    msg,
-                    logSessionId,
-                    workspace?.Id,
-                    workspace?.Tip?.Index,
-                    workspace?.Tip?.Hash
+                throw new ArgumentException(
+                    "The workspace chain must not have any renderers.",
+                    nameof(workspace)
                 );
-                foreach (var id in scope.Where(guid => guid != workspace?.Id))
+            }
```

Comment:
Can you clarify as to why `ExecuteActions()` care whether `workspace` has any renderer or not?

Reply:
There were condition in preloading logic, I just brought it from the previous code :)